### PR TITLE
ENH Add positive constraint to `L1`, `WeightedL1` and `L1_plus_L2`

### LINKED ISSUE #107: FEAT Add positivity constraint for `L1` and `L21` penalties

### PR DESCRIPTION
**Pull request by PABannier**
Add positivity constraints to these penalties and corresponding estimators.

Question: do we want it enabled for `WeightedL1`? It seems more consistent and is not tricky to do.

Closes #107, #132

### CODE REVIEW

**Comment by PABannier** [comment]
So far, it's working with `ws_strategy="fixpoint"`, there remains to write the correct distance to the subdifferential for all these penalties.

**Comment by PABannier** [comment]
Anyone to discuss with me about the impact of the positive constraint on the distance to the subdifferential. I try to do the maths, but I'm stuck... (poke @mathurinm @Klopfe @QB3 @Badr-MOUFAD )

**Comment by mathurinm** [comment]
For L1 + positivity constraint, when x is strictly negative, the penalty is + infty and the subdifferential is empty, so distance is infinite. When x is 0, the subdiff is ]-infty,1]. When x is > 0, the subdiff is the singleton 1

**Comment by PABannier** [comment]
Note to self: do same for L21, docstrings and implem might need compression

File: skglm/penalties/separable.py [hunk 1]
```diff
@@ -10,37 +10,49 @@
...
+        return ST(value, self.alpha * stepsize, self.positive)
 
     def subdiff_distance(self, w, grad, ws):
         """Compute distance of negative gradient to the subdifferential at w."""
         subdiff_dist = np.zeros_like(grad)
         for idx, j in enumerate(ws):
-            if w[j] == 0:
-                # distance of - grad_j to  [-alpha, alpha]
-                subdiff_dist[idx] = max(0, np.abs(grad[idx]) - self.alpha)
+            if self.positive:
```
Comment (Badr-MOUFAD) [inline skglm/penalties/separable.py:39]: Suggestion that I think would enhance the readability of this part

```python
# case of positivity constraints on coefficients 
if positive:
    # loop over features to compute subdiff distance
    ...
    ...
    return subdiff distance

# usual case
...
....
...
```

WDTY? (@mathurinm any thoughts?)

File: skglm/utils.py [hunk 2]
```diff
@@ -89,11 +89,11 @@ def compiled_clone(instance, to_float32=False):
 
 
 @njit
-def ST(x, u):
+def ST(x, u, positive):
```
Comment (Badr-MOUFAD) [inline skglm/utils.py:92]: We better make `positive` an optional argument (`positive=False`)?

- It won't impact other parts of the code that uses `ST`
- It's not intuitive to pass in a 3rd argument to ST

File: skglm/estimators.py [hunk 3]
```diff
@@ -492,8 +500,8 @@ class WeightedLasso(LinearModel, RegressorMixin):
     """
 
     def __init__(self, alpha=1., weights=None, max_iter=50, max_epochs=50_000, p0=10,
```
Comment (Badr-MOUFAD) [inline skglm/estimators.py:502]: **Suggestion:**

It appears that ``Lasso``, and `WeightedLasso` share a lot of similarities, noticeably in handling the case of weights=None for ``WeighedLasso``.

We shall think of merging both of them into one estimator.

WDYT @PABannier @mathurinm ?

**Review by Badr-MOUFAD — COMMENTED** [review]
hat off @PABannier for this work!

Some remarks and some suggestions :arrow_down:

File: skglm/estimators.py (hunk 3 above)
Comment (PABannier) [reply skglm/estimators.py:502]: Indeed! actually we could even have `ElasticNet` be the base class for `Lasso` (l1_ratio = 1), and `Lasso` the base class for `WeightedLasso`, as in `scikit-learn`. Anyway, if you find it relevant, open a new issue and let's discuss it there ;)

File: skglm/penalties/separable.py (hunk 1 above)
Comment (PABannier) [reply skglm/penalties/separable.py:39]: i find it redundant. I think it's already quite readable. @mathurinm  ?

Comment (PABannier) [reply skglm/penalties/separable.py:39]: Apart from that, anything else? (any suggestions @mathurinm ?)

**Comment by PABannier** [comment]
@mathurinm @Badr-MOUFAD should be mergeable

File: skglm/penalties/separable.py [hunk 4]
```diff
@@ -82,25 +95,37 @@ def value(self, w):
...
-                    0, np.abs(grad[idx]) - self.alpha * self.l1_ratio)
+            if self.positive:
+                if w[j] < 0:
+                    subdiff_dist[idx] = np.inf
+                elif w[j] == 0:
+                    # distance of -grad_j to (-infty, alpha * l1_ratio]
+                    subdiff_dist[idx] = max(0, - grad[idx] - self.alpha * self.l1_ratio)
+                else:
+                    # distance of -grad_j to alpha * {l1_ratio + (1 - l1_ratio) * w[j]}
+                    subdiff_dist[idx] = np.abs(
```
[the remark this answers is not in the document]
Comment (PABannier) [reply skglm/penalties/separable.py:114]: I don't find it particularly hard to read. But one thing that could be improved is to factor the `alpha` for less redundancy.

**Review by Badr-MOUFAD — APPROVED** [review]
LGTM! Many thanks @PABannier.

We should consider adding this feat to `Lasso`,  `WeightedLasso` and `ElasticNet` estimators

**Pull request merged**